lib/src/command/add.dart

```diff
@@ -0,0 +1,326 @@
+    final constraint = splitPackage.length == 2
+        ? VersionConstraint.parse(splitPackage[1])
+        : null;
+
+    /// Determine the relevant [packageRange] and [pubspecInformation] depending
+    /// on the type of package.
+    if (hasGitOptions) {
+      dynamic git;
+
+      /// Process the git options to return the simplest representation to be
+      /// added to the pubspec.
+      if (gitRef == null && gitPath == null) {
+        git = gitUrl;
+      } else {
+        git = {'url': gitUrl, 'ref': gitRef, 'path': gitPath};
+        git.removeWhere((key, value) => value == null);
+      }
+
+      packageRange = cache.sources['git']
+          .parseRef(packageName, git)
+          .withConstraint(VersionConstraint.any);
+      pubspecInformation = {'git': git};
+    } else if (path != null) {
+      packageRange = cache.sources['path']
+          .parseRef(packageName, path, containingPath: entrypoint.pubspecPath)
+          .withConstraint(VersionConstraint.any);
+      pubspecInformation = {'path': path};
+    } else if (sdk != null) {
+      packageRange = cache.sources['sdk']
+          .parseRef(packageName, sdk)
+          .withConstraint(constraint ?? VersionConstraint.any);
+      pubspecInformation = {'sdk': sdk};
+      if (constraint != null) {
+        pubspecInformation['version'] = constraint.toString();
+      }
+    } else {
+      final hostInfo =
+          hasHostOptions ? {'url': hostUrl, 'name': packageName} : null;
+
+      if (hostInfo == null) {
+        pubspecInformation = constraint?.toString();
+      } else if (constraint == null) {
+        pubspecInformation = {'hosted': hostInfo};
+      } else {
+        pubspecInformation = {
+          'hosted': hostInfo,
+          'version': constraint.toString()
+        };
+      }
+
+      packageRange = PackageRange(packageName, cache.sources['hosted'],
+          constraint ?? VersionConstraint.any, hostInfo ?? packageName);
+    }
+
+    return Pair(packageRange, pubspecInformation);
+  }
+
+  /// Writes the changes to the pubspec file.
+  void _updatePubspec(
+      PackageId resultPackage, Pair<PackageRange, dynamic> packageInformation) {
+    ArgumentError.checkNotNull(resultPackage, 'resultPackage');
+    ArgumentError.checkNotNull(packageInformation, 'pubspecInformation');
+
+    final package = packageInformation.first;
+    var pubspecInformation = packageInformation.last;
+
+    if ((sdk != null || hasHostOptions) &&
+        pubspecInformation is Map &&
+        pubspecInformation['version'] == null) {
+      /// We cannot simply assign the value of version since it is likely that
+      /// [pubspecInformation] takes on the type
+      /// [Map<String, Map<String, String>>]
+      pubspecInformation = {
+        ...pubspecInformation,
+        'version': '^${resultPackage.version}'
+      };
+    }
+
+    final dependencyKey = isDevelopment ? 'dev_dependencies' : 'dependencies';
```

Comment:
Somehow it would seem nicer to me if `isDevelopment` was passed as a parameter to `_updatePubspec`, that makes it easier to understand in isolation.